### DISCUSSION TITLE
Why multiple on_config is possible with rules but not directly in the target scope ?

### DISCUSSION THREAD

**Discussion by mactul** [xmake-io/xmake#5264] [Q&A]
Hello !
I'm trying to convert a big project to xmake and I struggle with all the scopes of xmake.
I have multiple scripts to run before the build that generate various .h/.c

I quickly realized that I can't call more than one on_config in the target scope.
For example this will just print "hello"
```lua
target("test")
    set_kind("phony")

    on_config(function (target) 
        print("hi")
    end)

    on_config(function (target) 
        print("hello")
    end)
```

But I just saw that on the other end, this displays "bar", "foo" and "hello"
```lua
rule("foo")
    on_config(function (target) 
        print("foo")
    end)

rule("bar")
    on_config(function (target)
        print("bar")
    end)


target("test")
    set_kind("phony")

    on_config(function (target) 
        print("hello")
    end)

    add_rules("foo", "bar")
```

Can someone explain to me what's the fundamental difference between a rule and the target itself that allows this behavior ?

Thank you in advance,
Mactul

**Comment by waruqi** [comment]
> I quickly realized that I can't call more than one on_config in the target scope.
For example this will just print "hello"

you can only set only one on_config in target. if you want to more, please use rules

**Comment by mactul** [comment]
Thank you for your reply,
Yes, I've observed that.
But I thought that the rules were just macros of sorts, used to factorize configurations that we often do.
Obviously, they're much more than that. I'm having trouble understanding what they actually represent.

But that's okay, at least I now know that they allow me to bypass some restrictions of the target scope.
